transaction.go

```diff
@@ -0,0 +1,489 @@
+		if err != nil {
+			return false
+		}
+		// RCD should match the input
+		if *expAddr != *s.FCTInputs[i].Address {
+			return false
+		}
+
+		if !s.Signatures[i].Validate(msg) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (s *FactoidTransaction) TotalFCTInputs() (total uint64) {
+	return FactoidTransactionIOs(s.FCTInputs).TotalAmount()
+}
+
+func (s *FactoidTransaction) TotalFCTOutputs() (total uint64) {
+	return FactoidTransactionIOs(s.FCTOutputs).TotalAmount()
+}
+
+// TotalECOutput is delimated in factoishis
+func (s *FactoidTransaction) TotalECOutput() (total uint64) {
+	return FactoidTransactionIOs(s.ECOutputs).TotalAmount()
+}
+
+func (s FactoidTransactionIOs) TotalAmount() (total uint64) {
+	for _, io := range s {
+		total += io.Amount
+	}
+	return
+}
+
+func (s FactoidTransactionSignature) Validate(msg Bytes) bool {
+	return s.ReedeemCondition.Validate(msg, s.SignatureBlock)
+}
+
+// Get queries factomd for the entry corresponding to f.TransactionID, which must be not
+// nil. After a successful call all inputs, outputs, and the header will be populated
+func (f *FactoidTransaction) Get(c *Client) error {
+	// TODO: Test this functionality
+	// If the TransactionID is nil then we have nothing to query for.
+	if f.TransactionID == nil {
+		return fmt.Errorf("txid is nil")
+	}
+	// If the Transaction is already populated then there is nothing to do. If
+	// the Hash is nil, we cannot populate it anyway.
+	if f.IsPopulated() {
+		return nil
+	}
+	requestTxId := f.TransactionID
+
+	params := struct {
+		Hash *Bytes32 `json:"hash"`
+	}{Hash: f.TransactionID}
+	var result struct {
+		Data Bytes `json:"data"`
+	}
+	if err := c.FactomdRequest("raw-data", params, &result); err != nil {
+		return err
+	}
+
+	if err := f.UnmarshalBinary(result.Data); err != nil {
+		return err
+	}
+
+	txid, err := f.ComputeTransactionID()
+	if err != nil {
+		return err
+	}
+	if txid != *requestTxId {
+		return fmt.Errorf("invalid txid")
+	}
+
+	return nil
+}
+
+// ComputeTransactionID computes the txid for a given transaction. The txid is the sha256 of
+// the ledger fields in a factoid transaction. The ledger fields exclude the signature block of
+// the transaction
+func (f *FactoidTransaction) ComputeTransactionID() (Bytes32, error) {
+	data, err := f.MarshalLedgerBinary()
+	if err != nil {
+		return Bytes32{}, err
+	}
+
+	txid := Bytes32(sha256.Sum256(data))
+	return txid, nil
+}
+
+// ComputeFullHash computes the fullhash for a given transaction. The fullhash is the sha256 of all
+// the fields in a factoid transaction.
+func (f *FactoidTransaction) ComputeFullHash() (*Bytes32, error) {
+	data, err := f.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	txid := Bytes32(sha256.Sum256(data))
+	return &txid, nil
+}
+
+// MarshalLedgerBinary marshals the transaction ledger fields to their binary representation.
+// This excludes the signature blocks
+func (f *FactoidTransaction) MarshalLedgerBinary() ([]byte, error) {
```

Comment:
Since computing hashes, keymrs, etc, should now be moved to the unmarshalBinary function where it can be done initially more efficiently on incoming data, the marshal functions do not need to be split up to accommodate the computation of keymrs and hashes after the fact. 

I'm not sure it adds all that much clarity to have these things broken up anymore and since the other marshaling functions now likely won't be used in any other context, they won't be reused anywhere. The other advantage to putting them in a single function is that it makes the allocations very explicit and clear.

Reply:
Currently, it is used in the Valid() function. So if you build a tx with this library, you would need this function to validate the signatures.

I did however remove the use of this in the Decode to compute the txid, so the decode follows the unmarshal idea of not reallocating bytes.

Reply:
We decided to keep it because users can build txs, and this is needed to verify signatures.